doc/horizontal_autoscaling.md

```diff
@@ -0,0 +1,173 @@
+# Design Doc: Horizontal Autoscaling
+
+## Background
+
+A PaddlePaddle training job contains several trainer instances,
+several parameter server instances, and one master instance. We would
+like to automatically scale the number of training instances and the
+number of parameter server instances to fully utilize the cluster's
+computation resources.
+
+Currently, we will only support horizontal trainer
+autoscaling. Parameter server horizontal autoscaling will be supported
+in the near future. This design doc considers both of them.
+
+[Horizontal Pod Autoscaling (HPA)](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/) is
+an autoscaling solution provided by Kuberentes, but it's not suitable
+for the training job autoscaling for the following reasons:
+
+- The training job autoscaling tries to fairly distribute the
+  computation resources to different training jobs, in order to
+  optimize the computation resource utilization of the **entire**
+  cluster. The HPA is trying to improve the quality of service of a
+  **single** service. The training job autoscaling requires the
+  controller to have a global view of all the available computation
+  resources and all the training jobs, but HPA does not have the
+  global view.
+
+- HPA is designed to automatically scale a homogeneous set of Pods,
+  but we need to scale a heterogeneous set of Pods (the trainer Pods
+  and the parameter server Pods): because the required number of
+  parameter servers is correlated to the required number of trainers,
+  we need to scale them together.
+
+We need to develop our own solution for autoscaling.
+
+## Solution
+
+We will build
+an [Operator](https://coreos.com/blog/introducing-operators.html) that
+do the autoscaling. To be more precise, we will create
+a
+[custom Kubernetes Resource](https://kubernetes.io/docs/concepts/api-extension/custom-resources/) and
+a
+[custom Kuberentes Controller](https://resources.coreos.com/youtube-coreos-fest-2017/writing-a-custom-controller-extending-the-functionality-of-your-cluster). Kuberentes
+is designed to be flexible so adding custom Resources and custom
+Controllers into our cluster does not require modifying the Kubernetes
+source code.
+
+### Training Job Resource
+
+Just
+like
+[Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) is
+a resource that describes a deployment. We will have a training
+job
+[Custom Resource](https://kubernetes.io/docs/concepts/api-extension/custom-resources/) that
+describes the training job.
+
+A pseudo resource declaration (`training_job.yaml`) is as follows:
+
+```yaml
+apiVersion: cloud.paddlepaddle.org/v1
+kind: TrainingJob
+metadata:
+  name: job-1
+spec:
+  trainer:
+    entrypoint: "python train.py"
+    workspace: "/home/job-1/"
+    min-replica: 3
```

Comment:
Maybe we also need a CPU/GPU threshold, and the controller will auto-scaling while the CPU/GPU util is upper than the threshold.

Reply:
@Yancey1989 Great idea! I think the CPU / GPU / memory / network usage for the trainer is mostly constant in data parallelism. However the usage could be a good indicator of auto-scaling the trainer when doing model parallel, or the parameter server.

It seems a little far from our current state, since we don't support model parallel and scaling parameter server yet. Maybe we can discuss it in the future?

Reply:
Sure, we will discuss this in the future. And I just want to confirm what's the triggering condition of executing `scale`, for my understanding,  maybe there is two way to trigger the `scale`:
1. Use CPU / GPU / memory ... usage as the threshold.
  As above, we will discuss this in the future.
1. Use global resource usage as the threshold.
  If there is any idling resource in the cluster, training job will scale itself. But I think maybe this feature also need depends on **priority queue**, so that Controller will know which job will be reduced and which job will be increased.

Just my personal question, maybe I have some mistake.

Reply:
> Use CPU / GPU / memory ... usage as the threshold.
As above, we will discuss this in the future.

Yes, we can use these metrics for scaling trainer in model parallel, and scaling pserver. But I don't we should use these metrics for scaling training in data parallel. Because the CPU / GPU / memory / network usage for the trainer is mostly constant in data parallelism.

> Use global resource usage as the threshold.
If there is any idling resource in the cluster, training job will scale itself. But I think maybe this feature also need depends on priority queue, so that Controller will know which job will be reduced and which job will be increased.


Is this algorithm similar to the priority queue (priority according to the fulfillment score) in your thoughts? https://github.com/helinwang/cloud/blob/25d38d97901e309ef14ab0444d2630a6d98a9eb9/doc/horizontal_autoscaling.md#scaling-algorithm

Reply:
Thanks @helinwang , I think it's enough for currently, and [Capacity Scheduler](https://hadoop.apache.org/docs/r2.7.4/hadoop-yarn/hadoop-yarn-site/CapacityScheduler.html) in Yarn is a good example for the **priority queue**, we can also discuss this in the future :)

Reply:
@Yancey1989 thanks for the reference!